### PR TITLE
Add SVG validation

### LINKED ISSUE #1397: Validate that `svgIcon` is indeed an SVG
We are currently missing validation of icons being SVGs - the spec says that only svgs are supported, but we do not have validation to enforce that.

### PR DESCRIPTION
**Pull request by FrederikBolding**
Adds validation of the snap SVG icon.

Fixes #1397

### CODE REVIEW

**Comment by ritave** [comment]
Is this necessary? I would be ok with accepting .pngs and .jpgs, etc.

URLs and especially dataurls are not safe, but I don't see why not all local images.

**Comment by FrederikBolding** [comment]
> Is this necessary? I would be ok with accepting .pngs and .jpgs, etc.
> 
> URLs and especially dataurls are not safe, but I don't see why not all local images.

Hmm, the spec specifically says SVGs and everything is currently built in a way that only supports SVG. However, there is no check to determine if an icon is actually an SVG, so weird errors may occur when using non SVGs.

If we want to support non-SVGs we would have to spend some time getting that working across the platform and in the extension. We can do that if we want to, but I'm not sure if there are more security concerns with loading arbitrary files instead of SVG? I'm not sure why SVG was chosen originally.

**Comment by ritave** [comment]
> I'm not sure why SVG was chosen originally.
IIRC for ease of implementation

> so weird errors may occur when using non SVGs.
The end is `` react component, so it should work out of the box, the only places it might break is when we're operating on the npm package or manifest. I think people were already using pngs without problems at this point?

In any case, doesn't need to be before audit, but it could be.

**Comment by FrederikBolding** [comment]
> The end is `` React component, so it should work out of the box, the only places it might break is when we're operating on the npm package or manifest. I think people were already using pngs without problems at this point?

This is not the case, we expect SVGs in the extension and format the `src` property as such: https://github.com/MetaMask/metamask-extension/blob/832ce634fd4b2b09b3a33911852fd851d82a373e/ui/selectors/selectors.js#L634-L636

There may be other cases such as this, currently anything binary will probably break.

File: packages/snaps-utils/src/icon.test.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+      `The specified SVG icon exceeds the maximum size of ${SVG_MAX_BYTE_SIZE_TEXT}.`,
+    );
+  });
+
+  it('asserts that the file is a valid SVG', () => {
+    const icon = new VirtualFile({
+      value: stringToBytes('foo'),
+      path: 'foo.svg',
+    });
+    expect(() => assertIsSnapIcon(icon)).toThrow(
```
[the remark this answers is not in the document]
Comment (FrederikBolding) [reply packages/snaps-utils/src/icon.test.ts:36]: Yeah, that's a good idea. I'll wait till the discussion with Olaf wraps up though to not waste further time on this if we don't want to pursue it.

[the remark this answers is not in the document]
Comment (FrederikBolding) [reply packages/snaps-utils/src/icon.test.ts:37]: Done!

**Comment by FrederikBolding** [comment]
@SocketSecurity ignore-all

File: packages/snaps-utils/src/icon.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+import { VirtualFile } from './virtual-file';
+
+export const SVG_MAX_BYTE_SIZE = 100_000;
+export const SVG_MAX_BYTE_SIZE_TEXT = `${Math.floor(
+  SVG_MAX_BYTE_SIZE / 1000,
+)}kb`;
+
+export const assertIsSnapIcon = (icon: VirtualFile) => {
+  assert(icon.path.endsWith('.svg'), 'Expected snap icon to end in .svg');
```
Comment (Mrtenz) [inline packages/snaps-utils/src/icon.ts:12]:
```suggestion
  assert(icon.path.endsWith('.svg'), 'Expected snap icon to end in ".svg".');
```

File: packages/snaps-utils/src/icon.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+  assert(
+    Buffer.byteLength(icon.value, 'utf8') <= SVG_MAX_BYTE_SIZE,
+    `The specified SVG icon exceeds the maximum size of ${SVG_MAX_BYTE_SIZE_TEXT}.`,
+  );
+
+  assert(isSvg(icon.toString()), 'Snap icon must be a valid SVG');
```
Comment (Mrtenz) [inline packages/snaps-utils/src/icon.ts:19]:
```suggestion
  assert(isSvg(icon.toString()), 'Snap icon must be a valid SVG.');
```

**Review by Mrtenz — CHANGES_REQUESTED** [review]

**Review by Mrtenz — DISMISSED** [review]

**Review by Mrtenz — APPROVED** [review]

**Pull request merged**